### DISCUSSION TITLE
outcast affiliation without effect in moderated MUC rooms

### DISCUSSION THREAD

**Discussion by Dunedan** [processone/ejabberd#4175] [Questions] [answered]
I have a strange problem with the "outcast" affiliation not being honored for moderated MUC rooms (using Ejabberd 23.01 on Debian). Is this intentional or possibly a bug?

Here are the options of a test room I just set up. For now the room is unmoderated.

```
$ ejabberdctl get_room_options test_muc_room conference.localhost
title
description
allow_change_subj       false
allow_query_users       true
allow_private_messages  true
allow_private_messages_from_visitors    nobody
allow_visitor_status    true
allow_visitor_nickchange        true
public  true
public_list     true
persistent      true
moderated       false
captcha_protected       false
members_by_default      false
members_only    false
allow_user_invites      true
allow_subscription      false
password_protected      false
password
anonymous       true
presence_broadcast      [moderator,participant,visitor]
allow_voice_requests    false
voice_request_min_interval      1800
max_users       200
logging true
vcard
vcard_xupdate   undefined
captcha_whitelist       {0,nil}
mam     false
pubsub
enable_hats     false
lang    en
```

Affiliations before outcasting the test user:

```
$ ejabberdctl get_room_affiliations test_muc_room conference.localhost
admin   localhost       owner
test_user       localhost       member
```

So far everything is fine. test_user can join the MUC room and chat, just as intended.

When I set the affiliation to "outcast" for the test user now, everything still behaves as expected: the user gets kicked from the room and can't rejoin. The room affiliations now look like:

```
$ ejabberdctl get_room_affiliations test_muc_room conference.localhost
admin   localhost       owner
test_user       localhost       outcast test
```

However, when I change the room to be moderated now, the user can suddently rejoin and chat, even though the affiliation is still in place. It doesn't matter if I change the moderated option before or after outcasting the user, the behavior is the same.

```
$ ejabberdctl get_room_options test_muc_room conference.localhost
title
description
allow_change_subj       false
allow_query_users       true
allow_private_messages  true
allow_private_messages_from_visitors    nobody
allow_visitor_status    true
allow_visitor_nickchange        true
public  true
public_list     true
persistent      true
moderated       true
captcha_protected       false
members_by_default      false
members_only    false
allow_user_invites      true
allow_subscription      false
password_protected      false
password
anonymous       true
presence_broadcast      [moderator,participant,visitor]
allow_voice_requests    false
voice_request_min_interval      1800
max_users       200
logging true
vcard
vcard_xupdate   undefined
captcha_whitelist       {0,nil}
mam     false
pubsub
enable_hats     false
lang    en

$ ejabberdctl get_room_affiliations test_muc_room conference.localhost
admin   localhost       owner
test_user       localhost       outcast test
```

Why does this happen? Is this supposed to be this way and if so why?

**Comment by licaon-kter** [comment]
`bookworm-backports` has 23.10, same issue?

**Reply by Dunedan** [reply]
With ejabberd 23.10 from `bookworm-backports` I can't reproduce the issue and outcasting works as I'd expect. :+1:
Is there by any chance a specific commit which fixed that, you can point me to?

**Reply by licaon-kter** [reply]
maybe https://github.com/processone/ejabberd/commit/dfe4884d16e640e7065c2ada293aaa5b6a2b9e46

**Reply by Dunedan** [reply, marked as answer]
I wonder if it might got fixed by #4010 instead.

**Reply by Neustradamus** [reply]
@Dunedan: Have you upgraded to master, it has been solved?

**Reply by Dunedan** [reply]
As mentioned above I tried it with ejabberd 23.10 and the issue was fixed there. I didn't try any more recent code.
